ports/raspberrypi/common-hal/alarm/SleepMemory.c

```diff
@@ -0,0 +1,63 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2021 Lucian Copeland for Adafruit Industries
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <string.h>
+
+#include "py/runtime.h"
+#include "common-hal/alarm/SleepMemory.h"
+
+// Data storage for singleton instance of SleepMemory.
+// Might be RTC_SLOW_MEM or RTC_FAST_MEM, depending on setting of CONFIG_ESP32S2_RTCDATA_IN_FAST_MEM.
+// static RTC_DATA_ATTR uint8_t _sleep_mem[SLEEP_MEMORY_LENGTH];
+
+void alarm_sleep_memory_reset(void) {
+    // ESP-IDF build system takes care of doing esp_sleep_pd_config() or the equivalentwith
+    // the correct settings, depending on which RTC mem we are using.
+    // https://docs.espressif.com/projects/esp-idf/en/latest/esp32s2/api-reference/system/sleep_modes.html#power-down-of-rtc-peripherals-and-memories
```

Comment:
These comments pertain to ESP-IDF, so remove them for RP2040. If `SleepMemory` is completely not implemented for RP2040, I think that `alarm.sleep_memory` should probably raise `NotImplementedError`, instead of just being zero length.

Reply:
My bad, I missed that this block wasn't implemented. I'll put in a NotImplementedError for now, but I'll revisit this in a future PR to probably try and move the whole RP2040 system over to not using the watchdog reset, and implementing "sleep memory" in the heap space somehow.

Reply:
I'm wondering if it's worth merging this now if it's going to change in the future. What is the current consumption for light and deep sleep now. If they are high maybe we should leave this unimplemented now on RP2040 because it's not worth it.

Reply:
@dhalbert the only effect of the change on the RP2040 would be to enable Sleep Memory. The global change would be to remove "NORETURN" keyword from the API, which I think might be better tracked separately. Using PinAlarm on the Pico improves life on a 1000mAh battery from a day and a half (27mA) to about a month (1.4mA) on a board that isn't optimized for power, so I'm not sure why we'd want to toss the work on this PR.